crates/bevy_diagnostic/src/system_information_diagnostics_plugin.rs

```diff
@@ -121,76 +124,90 @@ pub mod internal {
-        mut last_refresh: Local<Option<Instant>>,
-    ) {
-        let sysinfo = sysinfo.get_or_insert_with(|| {
-            Arc::new(Mutex::new(System::new_with_specifics(
-                RefreshKind::nothing()
-                    .with_cpu(CpuRefreshKind::nothing().with_cpu_usage())
-                    .with_memory(MemoryRefreshKind::everything()),
-            )))
-        });
-
-        let last_refresh = last_refresh.get_or_insert_with(Instant::now);
-
-        let thread_pool = AsyncComputeTaskPool::get();
-
-        // Only queue a new system refresh task when necessary
-        // Queuing earlier than that will not give new data
-        if last_refresh.elapsed() > sysinfo::MINIMUM_CPU_UPDATE_INTERVAL
-            // These tasks don't yield and will take up all of the task pool's
-            // threads if we don't limit their amount.
-            && tasks.tasks.len() * 2 < available_parallelism()
-        {
-            let sys = Arc::clone(sysinfo);
-            let task = thread_pool.spawn(async move {
-                let mut sys = sys.lock().unwrap();
-                let pid = sysinfo::get_current_pid().expect("Failed to get current process ID");
-                sys.refresh_processes(sysinfo::ProcessesToUpdate::Some(&[pid]), true);
-
-                sys.refresh_cpu_specifics(CpuRefreshKind::nothing().with_cpu_usage());
-                sys.refresh_memory();
-                let system_cpu_usage = sys.global_cpu_usage().into();
-                let total_mem = sys.total_memory() as f64;
-                let used_mem = sys.used_memory() as f64;
-                let system_mem_usage = used_mem / total_mem * 100.0;
-
-                let process_mem_usage = sys
-                    .process(pid)
-                    .map(|p| p.memory() as f64 * BYTES_TO_GIB)
-                    .unwrap_or(0.0);
-
-                let process_cpu_usage = sys
-                    .process(pid)
-                    .map(|p| p.cpu_usage() as f64 / sys.cpus().len() as f64)
-                    .unwrap_or(0.0);
-
-                SysinfoRefreshData {
-                    system_cpu_usage,
-                    system_mem_usage,
-                    process_cpu_usage,
-                    process_mem_usage,
+    #[derive(Default, Resource)]
+    struct DiagnosticJobData {
+        data_rx: Option<Receiver<SysinfoRefreshData>>,
+    }
+
+    impl Future for DiagnosticJob {
+        type Output = ();
+
+        fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+            cx.waker().wake_by_ref();
+
+            let run_diagnostic = self
+                .last_refresh
+                .map(|last_refresh| last_refresh.elapsed() > sysinfo::MINIMUM_CPU_UPDATE_INTERVAL)
+                .unwrap_or(true);
+
+            if run_diagnostic {
+                {
+                    let mut sys = self.sysinfo.lock().unwrap();
+                    let data = get_diagnostic_data(&mut sys);
+                    _ = self.data_rx.send(data);
```

Comment:
I think this may have been a mistake in the sketch from my example. Since `send` returns a future, it needs to be awaited rather than dropped.

This should be `self.data_rx.send(data).await`.

Reply:
`send` doesn't return a `Future` (using `crossbeam-channel` here). and also, how would you await a`Future` inside the `impl Future`...

Reply:
Ah, using `crossbeam-channel` this will block the executor when full. You need to use `try_send` or an async channel.

We could also write this task as an async block to make using a channel easier.

```rust
async move {
    loop {
        // all the polling and stuff
        data_rx.send(data).await;
        futures::yeild_now().await;
    }
}
```

Should be approximately the same.